feat: Cure effects food property

RealRONiN opened the pull request:
Honey bottles can cure any effect same as milk bukkets, it can also heal a player and add additional potion effects if any specified.
IRL Honey has numerous benefits, it'd make sense if it serves more benefits than just the saturation effect.

RealRONiN committed:
feat: honey bottle cures and heals

RealRONiN commented:
I'm not sure what to do with this, https://github.com/PurpurMC/Purpur/blob/c27e67a4a830c1c9c003a4da80226a5f98bf6d95/patches/server/0309-Honey-bottle-cures-and-heals.patch#L32

BillyGalbreath commented on patches/server/0309-Honey-bottle-cures-and-heals.patch:
```diff
@@ -0,0 +1,141 @@
++    public float honeyBottleRegenAmount = 1.5f;
++    public boolean honeyBottleAddsPositiveEffects = true;
++    public int honeyBottlePotionEffectsDuration = 200;
++    public List<MobEffect> honeyBottleCurableEffects = new ArrayList<>();
```
I suggest using the word remove instead of cure/curable in the name (`honeyBottleRemoveEffects` here, and `remove-effects` below). It goes with the `add` option better, imo.

BillyGalbreath commented on patches/server/0309-Honey-bottle-cures-and-heals.patch:
```diff
@@ -0,0 +1,141 @@
++    public boolean honeyBottleAddsPositiveEffects = true;
++    public int honeyBottlePotionEffectsDuration = 200;
++    public List<MobEffect> honeyBottleCurableEffects = new ArrayList<>();
++    public List<MobEffect> honeyBottleAddedEffects = new ArrayList<>();
```
If you change this to a Map<MobEffect,Integer> we can have per-effect durations and can get rid of the extra `honeyBottlePotionEffectsDuration` option.

BillyGalbreath commented on patches/server/0309-Honey-bottle-cures-and-heals.patch:
```diff
@@ -0,0 +1,141 @@
++            add("minecraft:wither");
++            add("minecraft:levitation");
++            add("minecraft:unluck");
++            add("minecraft:bad_omen");
```
Make this only contain vanilla behavior (only removes poison), then you can get rid of the extra `honeyBottleCures` option above.

BillyGalbreath commented on patches/server/0309-Honey-bottle-cures-and-heals.patch:
```diff
@@ -0,0 +1,141 @@
++            add("minecraft:strength");
++            add("minecraft:regeneration");
++            add("minecraft:saturation");
++            add("minecraft:glowing");
```
Make this only contain vanilla behaviors (empty section), then you can get rid of the extra `honeyBottleAddsPositiveEffects` option above.

BillyGalbreath commented on patches/server/0309-Honey-bottle-cures-and-heals.patch:
```diff
@@ -0,0 +1,141 @@
+@@ -29,6 +36,33 @@ public class HoneyBottleItem extends Item {
+ 
+         if (!world.isClientSide) {
+             user.removeEffect(MobEffects.POISON, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // Paper
```
You can get rid of this line entirely. Just be sure poison is in the default options. That way if someone wants they could actually remove it from being curable.

RealRONiN replied:
Moved it to defualt food properties

BillyGalbreath reviewed (changes requested):
I just woke up, so I hope I got it all in one go ^_^

RealRONiN committed:
refactor: cure effects food property

Removed previous honey bottle patches to include cure effect property in food properties so any food item can use it.

RealRONiN committed:
refactor: moved Paper's honey bottle cure poison to default food properties

RealRONiN commented:
Technically, milk bukket should be a food item, should I add it to `Foods` class and move MilkBukketItem class's properties to it?

RealRONiN committed:
revert: reverted some formattings and auto-imports

RealRONiN commented:
https://discord.com/channels/685683385313919172/720374890472931328/1028914991421726730

granny committed:
Merge branch 'ver/1.19.2' into RealRONiN/ver/1.19.2

granny committed:
rebase

granny committed:
deep clean patch

from 21608 characters to 9759; over 54% of characters were removed.

from 304 lines long to 160; 47% of lines were removed.

granny reviewed (approved):
PR is ready to merge. Waiting on Billy in case he wants to make any changes.

BillyGalbreath reviewed (changes requested):
This is _far_ from ready. I will rewrite it myself when I get back online.

granny commented:
Closing as it doesn't seem like there will be any progress done anytime soon.
